### PR TITLE
Add documentation for sparse ops

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
Summary:
For operators, the input / output types are still `NDArray`s instead of specific `RowSparseNDArray` or `CSRNDArray`. It's a bit difficult make this change. Instead I am adding sections in operator doc to specify what sparse operators are implemented. 

Reason:
Since `NDArray-or-Symbol` is used for all language bindings to generate their code, I want to avoid changing it for backward compatibility. Meanwhile `NDArray-or-Symbol` also determines the API doc generated. (See cpp-package/scripts/OpWrapperGenerator.py & python/mxnet/ndarray_doc.py.) Explicitly stating what sparse operators in description should be enough for users interested in sparse operators to figure out what's supported.  @reminisce @piiswrong thoughts? 

preview: 

http://ec2-54-187-32-207.us-west-2.compute.amazonaws.com/api/python/ndarray.html#mxnet.ndarray.elemwise_add

### CODE REVIEW

File: python/mxnet/ndarray/sparse_ndarray.py [hunk 1]
```diff
@@ -846,8 +846,8 @@ def _zeros_sparse_ndarray(stype, shape, ctx=None, dtype=None, aux_types=None, **
     dtype : str or numpy.dtype, optional
         An optional value type (default is `float32`)
     aux_types: list of numpy.dtype, optional
-        An optional type for the aux data for BaseSparseNDArray (default values depends
-        on the storage type)
+        An optional list of types for the aux data for RowSparseNDArray or CSRNDArray
```
Comment (reminisce) [inline python/mxnet/ndarray/sparse_ndarray.py]: Change the second `for` to `of`?

File: python/mxnet/ndarray/sparse_ndarray.py [hunk 2]
```diff
@@ -846,8 +846,8 @@ def _zeros_sparse_ndarray(stype, shape, ctx=None, dtype=None, aux_types=None, **
...
+        (default values depends on the storage type)
```
Comment (reminisce) [inline python/mxnet/ndarray/sparse_ndarray.py]: depends -> depend

File: python/mxnet/ndarray/utils.py [hunk 3]
```diff
@@ -39,8 +39,8 @@ def zeros(shape, ctx=None, dtype=None, stype=None, aux_types=None, **kwargs):
     stype: string, optional
         The storage type of the empty array, such as 'row_sparse', 'csr', etc
```
Comment (reminisce) [inline python/mxnet/ndarray/utils.py]: etc -> etc.

File: python/mxnet/ndarray/utils.py [hunk 4]
```diff
@@ -39,8 +39,8 @@ def zeros(shape, ctx=None, dtype=None, stype=None, aux_types=None, **kwargs):
...
     aux_types: list of numpy.dtype, optional
-        An optional type for the aux data for the BaseSparseNDArray (default values
-        depends on the storage type)
+        An optional list of types for the aux data for RowSparseNDArray or CSRNDArray
```
Comment (reminisce) [inline python/mxnet/ndarray/utils.py]: second `for` -> `of`?

File: python/mxnet/ndarray/utils.py [hunk 5]
```diff
@@ -39,8 +39,8 @@ def zeros(shape, ctx=None, dtype=None, stype=None, aux_types=None, **kwargs):
...
+        (default values depends on the storage type)
```
Comment (reminisce) [inline python/mxnet/ndarray/utils.py]: `depends` -> `depend`?

File: python/mxnet/ndarray/utils.py [hunk 6]
```diff
@@ -73,8 +73,8 @@ def empty(shape, ctx=None, dtype=None, stype=None, aux_types=None):
     stype : str, optional
         An optional storage type (default is `default`).
     aux_types: list of numpy.dtype, optional
-        An optional type for the aux data for the BaseSparseNDArray (default values depends
-        on the storage type)
+        An optional list of types for the aux data for RowSparseNDArray or CSRNDArray
```
Comment (reminisce) [inline python/mxnet/ndarray/utils.py]: second `for` -> `of`?

File: python/mxnet/ndarray/utils.py [hunk 7]
```diff
@@ -73,8 +73,8 @@ def empty(shape, ctx=None, dtype=None, stype=None, aux_types=None):
...
+        (default values depends on the storage type)
```
Comment (reminisce) [inline python/mxnet/ndarray/utils.py]: `depends` -> `depend`?

File: python/mxnet/ndarray/utils.py [hunk 8]
```diff
@@ -111,8 +111,8 @@ def array(source_array, ctx=None, dtype=None, aux_types=None):
         The data type of the output array. The default dtype is ``source_array.dtype``
         if `source_array` is an `NDArray`, `float32` otherwise.
     aux_types: list of numpy.dtype, optional
-        An optional type for the aux data for the BaseSparseNDArray (default values
-        depends on the storage type)
+        An optional list of types for the aux data for RowSparseNDArray or CSRNDArray
```
Comment (reminisce) [inline python/mxnet/ndarray/utils.py]: same here.

File: python/mxnet/ndarray/utils.py [hunk 9]
```diff
@@ -111,8 +111,8 @@ def array(source_array, ctx=None, dtype=None, aux_types=None):
...
+        (default values depends on the storage type)
```
Comment (reminisce) [inline python/mxnet/ndarray/utils.py]: `depend`?

File: src/operator/tensor/cast_storage.cc [hunk 10]
```diff
@@ -32,6 +32,39 @@ namespace op {
 DMLC_REGISTER_PARAMETER(CastStorageParam);
 NNVM_REGISTER_OP(cast_storage)
 .describe(R"code(Casts tensor storage type to the new type.
+
+When an NDArray with default storage type is casted to csr or row_sparse storage,
```
Comment (reminisce) [inline src/operator/tensor/cast_storage.cc]: `casted` was the past tense of `cast` in old English. It's `cast` nowadays.
https://english.stackexchange.com/questions/94565/can-casted-be-the-past-tense-of-cast

File: src/operator/tensor/cast_storage.cc [hunk 11]
```diff
@@ -32,6 +32,39 @@ namespace op {
...
+the result is compact, which means:
+
+- for csr, zero values will not be retained
+- for row_sparse, row slices with zeros will not be retained
```
Comment (reminisce) [inline src/operator/tensor/cast_storage.cc]: `row slices with zeros` does not sound accurate. Is `row slices of all zeros` better?

**Comment by eric-haibin-lin** [comment]
Ping @cjolivier01 since binary/unary operators also need descriptions on storage types supported

File: src/operator/tensor/cast_storage.cc (hunk 10 above)
Comment (eric-haibin-lin) [inline src/operator/tensor/cast_storage.cc]: ha, my English is really bad

**Pull request merged**